### ISSUE TITLE
Can't get gdb to display C code

### ISSUE THREAD

**Issue by nbosio1001** [mit-pdos/xv6-riscv#171]
I am working on primes.c and I can't get my gdb environment setup so I can debug the code visually. I have the same .gdbinit.tmpl-riscv file and put $U/_primes\ in my Makefile. I am also running on a Raspberry Pi OS.

I run `make qemu-gdb` and then run `gdb-multiarch user/_primes`. After I run `lay next`, press enter and get the gdb environment in the picture.

![image](https://user-images.githubusercontent.com/49621575/223151316-4a4de42b-1939-47eb-8f3a-06b9d986eee2.png)

**Comment by meighti** [comment]
When you run make qemu-gdb, you are debugging the kernel, not user processes like user/_primes.

You might want to take a look at trampoline.S, which has a 'sret' instruction for returning to user mode. I think after 'sret' you won't be able to debug into user process and you will get back to gdb once you are back somewhere in the kernel, (probably via a syscal or timer interrupt).

In GDB, use "b trap.c:128" to break before jumping to userret. Then use "si" to step by instruction, until you reach to 'sret'. after that you will see that GDB sill not step from supervisor mode into user mode.

**Comment by chenzhiy2001** [comment]
> I think after 'sret' you won't be able to debug into user process and you will get back to gdb once you are back somewhere in the kernel

That's true if you use `gdb-multiarch`. If you want to single stepping from kernel to a user process, you can use `riscv64-unknown-elf-gdb` instead of `gdb-multiarch`. I guess they handle single stepping `sret` instructions differently.

I recommend you to compile  `riscv64-unknown-elf-gdb` yourself from [riscv-gnu-toolchain](https://github.com/riscv-collab/riscv-gnu-toolchain#installation-newlib).

Source: https://stackoverflow.com/a/69961102/24061604

**Comment by kaashoek** [comment]
Closing expired issue
